src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -3954,21 +3908,368 @@ ColocationGroupDeleteCommand(uint32 colocationId)
+	metadataSyncContext->workerOutsideTranConnections = workerOutsideTranConnections;
+}
+
+
+/*
+ * CreateMetadataSyncContext creates a context which contains worker connections
+ * and a MemoryContext to be used for all transactions used throughout the metadata
+ * sync.
+ *
+ * If commandAccumulationForTest is true, connections will not be established as
+ * caller's intent is to accumulate sync commands for test purposes.
+ */
+MetadataSyncContext *
+CreateMetadataSyncContext(List *nodeList, bool useCoordinatedTran,
+						  bool commandAccumulationForTest)
+{
+	/* should be alive over multiple transactions during the sync */
+	MemoryContext context = AllocSetContextCreate(TopMemoryContext,
+												  "metadata_sync_context",
+												  ALLOCSET_DEFAULT_SIZES);
+
+	MetadataSyncContext *metadataSyncContext = (MetadataSyncContext *) palloc0(
+		sizeof(MetadataSyncContext));
+
+	metadataSyncContext->context = context;
+
+	metadataSyncContext->useCoordinatedTran = useCoordinatedTran;
+
+	/* filter metadata syncable nodes from given node list */
+	SetMetadataSyncableNodesFromNodeList(metadataSyncContext, nodeList);
+
+	metadataSyncContext->commandAccumulationForTest = commandAccumulationForTest;
+	metadataSyncContext->accumulatedCommands = NIL;
+
+	/* do not initiate connections if caller aims testing */
+	if (metadataSyncContext->commandAccumulationForTest)
+	{
+		return metadataSyncContext;
+	}
+
+	SetMetadataSyncOutsideTransactionConnections(metadataSyncContext);
+
+	return metadataSyncContext;
+}
+
+
+/*
+ * ResetMetadataSyncMemoryContext resets memory context inside metadataSyncContext
+ * if we are not in test mode.
+ */
+void
+ResetMetadataSyncMemoryContext(MetadataSyncContext *metadataSyncContext)
+{
+	if (!metadataSyncContext->commandAccumulationForTest)
+	{
+		MemoryContextReset(metadataSyncContext->context);
+	}
+}
 
+
+/*
+ * SendCommandListToWorkerListViaMetadataContext sends the commands to the worker list
+ * with outside transactions inside outside connections at metadatacontext or
+ * with coordinated transaction via metadata connections. It decides the connection mode
+ * after checking the flag useCoordinatedTran inside metadatacontext.
+ */
+void
+SendCommandListToWorkerListViaMetadataContext(MetadataSyncContext *metadataSyncContext,
+											  List *commands)
+{
+	/*
+	 * do not send any command to workers if it is only for test purpose.
+	 * Accumulate commands into metadataSyncContext's accumulated command
+	 * list.
+	 */
+	if (metadataSyncContext->commandAccumulationForTest)
+	{
+		metadataSyncContext->accumulatedCommands =
+			list_concat(metadataSyncContext->accumulatedCommands,
+						commands);
+		return;
+	}
+
+	if (metadataSyncContext->useCoordinatedTran)
+	{
+		SendMetadataCommandListToWorkerListInCoordinatedTransaction(
+			metadataSyncContext->syncableWorkerNodeList,
+			CurrentUserName(),
+			commands);
+	}
+	else
+	{
+		SetMetadataSyncOutsideTransactionConnections(metadataSyncContext);
```

Comment:
We donot need after I refactor. We should exclusively claim outside connections. Otherwise, 2pc connection uses cached connection and does not release it until local transaction ends. Then upcoming outside transaction could not use already assigned connection.